src/asl.ts

```diff
@@ -5526,28 +5253,1768 @@ export namespace ASLGraph {
+              props,
+            },
+          ];
+        } else {
+          return Object.entries(parameters).flatMap(([key, param]) => {
+            const jsonPath = key.endsWith(".$");
+            return parametersAssignment(param, jsonPath, resultPath, [
+              ...props,
+              jsonPath ? key.substring(0, key.length - 2) : key,
+            ]);
+          });
+        }
+      }
+      if (!containsJsonPath) {
+        return [
+          {
+            type: "LiteralPropAssignment",
+            to: resultPath,
+            value: parameters,
+            props,
+          },
+        ];
+      }
+      if (typeof parameters === "string") {
+        if (parameters.startsWith("States.")) {
+          return extractVariableReferences(parameters).map((x) => ({
+            type: "Intrinsic",
+            from: x,
+            to: resultPath,
+            props,
+          }));
+        } else if (parameters.includes("?(") || parameters.includes(":")) {
+          return extractVariableReferences(parameters).map((x) => ({
+            type: "FilterPropAssignment",
+            from: x,
+            to: resultPath,
+            props,
+          }));
+        }
+      }
+      // if the value is json path, it will not be a boolean or string
+      return [
+        resultPath === "[[task]]" || resultPath === "[[map]]"
+          ? {
+              type: "StateInputProps",
+              from: parameters as string,
+              props,
+              stateType: resultPath === "[[task]]" ? "Task" : "Map",
+            }
+          : {
+              type: "PropAssignment",
+              from: parameters as string,
+              to: resultPath,
+              props,
+            },
+      ];
+    }
+
+    /**
+     * Pull variable references out of intrinsic functions and json path filters.
+     *
+     * States.Array($.var) => $.var
+     * $.arr[?(['value'] == $.var)] => $.arr, $.var
+     */
+    function extractVariableReferences(jsonPath: string) {
+      return jsonPath.match(/(\$\.?[^,)]+)+/g) ?? [];
+    }
+  }
+
+  function getVariableNamesFromState(state: State): string[] {
+    const names = ((state) => {
+      if (
+        isTaskState(state) ||
+        isMapTaskState(state) ||
+        isParallelTaskState(state)
+      ) {
+        return [
+          state.ResultPath,
+          ...(state.Catch?.map((c) => c.ResultPath) ?? []),
+        ];
+      } else if (isPassState(state)) {
+        return [state.ResultPath];
+      } else if (
+        isSucceedState(state) ||
+        isFailState(state) ||
+        isChoiceState(state) ||
+        isWaitState(state)
+      ) {
+        return [];
+      }
+      return assertNever(state);
+    })(state);
+
+    return names.filter((n): n is string => !!n);
+  }
+
+  export const DefaultOptimizeOptions: OptimizeOptions = {
+    optimizeVariableAssignments: true,
+    removeUnreachableStates: true,
+    joinConsecutiveChoices: true,
+    removeNoOpStates: true,
```

Comment:
All on by default?

Reply:
Yes, user should not have to know about these unless they start to break or if we make more experimental ones.

Reply:
Does the JSON Path re-write worry you?